ocaml/xapi/xapi_vm_lifecycle.ml

```diff
@@ -478,36 +490,17 @@ let check_operation_error ~__context ~vmr ~vmgmr ~ref ~clone_suspended_vm_enable
 
   current_error
 
-let maybe_get_guest_metrics ~__context ~ref =
-  if Db.is_valid_ref __context ref
-  then Some (Db.VM_guest_metrics.get_record_internal ~__context ~self:ref)
-  else None
-
-let get_info ~__context ~self =
-  let all = Db.VM.get_record_internal ~__context ~self in
-  let gm = maybe_get_guest_metrics ~__context ~ref:(all.Db_actions.vM_guest_metrics) in
-  let clone_suspended_vm_enabled = Helpers.clone_suspended_vm_enabled ~__context in
-  let vdis_reset_and_caching = List.filter_map (fun vbd ->
-      try
-        let vdi = Db.VBD.get_VDI ~__context ~self:vbd in
-        let sm_config = Db.VDI.get_sm_config ~__context ~self:vdi in
-        Some ((assoc_opt "on_boot" sm_config = Some "reset"), (bool_of_assoc "caching" sm_config))
-      with _ -> None) all.Db_actions.vM_VBDs in
-  all, gm, clone_suspended_vm_enabled, vdis_reset_and_caching
-
 let get_operation_error ~__context ~self ~op ~strict =
-  let all, gm, clone_suspended_vm_enabled, vdis_reset_and_caching = get_info ~__context ~self in
-  check_operation_error __context all gm self clone_suspended_vm_enabled vdis_reset_and_caching op strict
+  check_operation_error ~__context ~ref:self ~op ~strict
 
 let assert_operation_valid ~__context ~self ~op ~strict =
   match get_operation_error ~__context ~self ~op ~strict with
   | None       -> ()
   | Some (a,b) -> raise (Api_errors.Server_error (a,b))
 
 let update_allowed_operations ~__context ~self =
-  let all, gm, clone_suspended_vm_enabled, vdis_reset_and_caching = get_info ~__context ~self in
   let check accu op =
-    match check_operation_error __context all gm self clone_suspended_vm_enabled vdis_reset_and_caching op true with
```

Comment:
We could actually restore the old performance of this function, without using `get_info` and introducing duplication:
```ocaml
let check_operation_error ~__context ~ref =
  let vmr = Db.VM.get_record_internal ~__context ~self:ref in
  let vmgmr = lazy (maybe_get_guest_metrics ~__context ~ref:(vmr.Db_actions.vM_guest_metrics)) in
  (fun ~op ~strict ->
      let x = lazy ... in
      ...
  )
```
Then, when we call it once only, the code remains the same. But when we call it in a loop with multiple operations, like in `update_allowed_operations`, we can do:
```ocaml
let check_operation_error ~__context ~ref = check_operation_error ~context ~ref in
loop (
  ... check_operation_error ~op ~strict:true
)
```